substrate/frame/sassafras/src/lib.rs

```diff
@@ -0,0 +1,1059 @@
+		fn on_initialize(block_num: BlockNumberFor<T>) -> Weight {
+			let claim = <frame_system::Pallet<T>>::digest()
+				.logs
+				.iter()
+				.filter_map(|item| item.pre_runtime_try_to::<SlotClaim>(&SASSAFRAS_ENGINE_ID))
+				.next()
+				.expect("Valid block must have a slot claim. qed");
+
+			CurrentSlot::<T>::put(claim.slot);
+
+			// As the slots may not be zero-based, we need to keep track of what is the
+			// slot used for the first block.
+			if frame_system::Pallet::<T>::block_number() == One::one() {
+				GenesisSlot::<T>::put(claim.slot);
+
+				// Deposit a log as this is the first block in first epoch.
+				let next_epoch = NextEpochDescriptor {
+					randomness: Self::next_randomness(),
+					authorities: Self::next_authorities().into_inner(),
+					config: None,
+				};
+				Self::deposit_next_epoch_descriptor_digest(next_epoch);
+			}
+
+			let randomness_output = claim
+				.vrf_signature
+				.outputs
+				.get(0)
+				.expect("Valid claim must have vrf signature; qed");
+			ClaimTemporaryData::<T>::put(randomness_output);
+
+			T::WeightInfo::on_initialize() +
+				T::EpochChangeTrigger::trigger::<T>(block_num).unwrap_or_default()
+		}
+
+		fn on_finalize(_: BlockNumberFor<T>) {
+			// At the end of the block, we can safely include the current slot randomness
+			// to the randomness accumulator. If we've determined that this block was the
+			// first in a new epoch, the changeover logic has already occurred at this point
+			// (i.e. `enact_epoch_change` has already been called).
+			let randomness_input = vrf::slot_claim_input(
+				&Self::randomness(),
+				CurrentSlot::<T>::get(),
+				EpochIndex::<T>::get(),
+			);
+			let randomness_output = ClaimTemporaryData::<T>::take()
+				.expect("Finalization is called after initialization; qed");
+			let randomness = randomness_output
+				.make_bytes::<RANDOMNESS_LENGTH>(RANDOMNESS_VRF_CONTEXT, &randomness_input);
+			Self::deposit_slot_randomness(&randomness);
+
+			// Check if we are in the epoch's second half.
+			// If so, start sorting the next epoch tickets.
+			let epoch_length = T::EpochLength::get();
+			let current_slot_idx = Self::current_slot_index();
+			if current_slot_idx >= epoch_length / 2 {
+				let mut metadata = TicketsMeta::<T>::get();
+				if metadata.unsorted_tickets_count != 0 {
+					let epoch_idx = EpochIndex::<T>::get() + 1;
+					let epoch_tag = (epoch_idx & 1) as u8;
+					let slots_left = epoch_length.checked_sub(current_slot_idx).unwrap_or(1);
+					Self::sort_tickets(
+						metadata
+							.unsorted_tickets_count
+							.div_ceil(SEGMENT_MAX_SIZE * slots_left as u32),
+						epoch_tag,
+						&mut metadata,
+					);
+					TicketsMeta::<T>::set(metadata);
+				}
+			}
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Submit next epoch tickets candidates.
+		///
+		/// The number of tickets allowed to be submitted in one call is equal to the epoch length.
+		// TODO: maybe we must be more restrictive?
```

Comment:
Yes, given the benchmarks someone submitting `MaxTicketsFor` would probably be too much. Let's keep it as is for now though, until we find an adequate bound.